usecase/providerd.go

```diff
@@ -73,26 +73,36 @@ func (g *providerDaemon) Start(ctx context.Context) <-chan []error {
 
 		for {
 			select {
-			case <-ctx.Done():
-				errs := make([]error, 0, len(emap)+1)
-				for err, count := range emap {
-					errs = append(errs, errors.WithMessagef(err, "%d times appeared", count))
-				}
-				errs = append(errs, ctx.Err())
-				ech <- errs
-				return
+			// ctx.Done() will be handled by sch
+			// case <-ctx.Done():
 			case e := <-pch:
+				if e == nil {
+					// prevent handling nil value on channel close
+					continue
+				}
 				glg.Errorf("pch %v", e)
-				glg.Error(e)
+				// count errors by cause
 				cause := errors.Cause(e)
 				_, ok := emap[cause]
 				if !ok {
-					emap[cause] = 0
+					emap[cause] = 1
+				} else {
+					emap[cause]++
 				}
-				emap[cause]++
-			case errs := <-sch:
-				glg.Errorf("sch %v", errs)
-				ech <- errs
+			case serrs := <-sch:
+				if serrs == nil {
```

Comment:
I think we should fix the nil problem in server.go line205 rather than here, we should not handle receiving nil value from channel as it is meaningless.

Reply:
why line are you referring?

It is about the race condition between the `pch` and `sch`. We cannot assume that they always closed in the expected order.